appmetrics/database.go

```diff
@@ -0,0 +1,161 @@
+
+type AppMetricValidationError struct {
+	Metric AppMetric
+	Errors []gojsonschema.ResultError
+}
+
+const (
+	// Events for testing the system
+	TestEvent1 AppMetricEventType = "go/test1"
+	TestEvent2 AppMetricEventType = "go/test2"
+
+	// status-react navigation events
+	NavigationNavigateToCofx AppMetricEventType = "navigation/navigate-to"
+)
+
+// EventSchemaMap Every event should have a schema attached
+var EventSchemaMap = map[AppMetricEventType]interface{}{
+	TestEvent1:               StringSchema,
+	TestEvent2:               StringSchema,
+	NavigationNavigateToCofx: NavigationNavigateToCofxSchema,
+}
+
+func NewDB(db *sql.DB) *Database {
+	return &Database{db: db}
+}
+
+// Database sql wrapper for operations with browser objects.
+type Database struct {
+	db *sql.DB
+}
+
+// Close closes database.
+func (db Database) Close() error {
+	return db.db.Close()
+}
+
+func jsonschemaErrorsToError(validationErrors []AppMetricValidationError) error {
+	var fieldErrors []string
+
+	for _, appMetricValidationError := range validationErrors {
+		metric := appMetricValidationError.Metric
+		errors := appMetricValidationError.Errors
+
+		var errorDesc string = "Error in event: " + string(metric.Event) + " - "
+		for _, e := range errors {
+			errorDesc = errorDesc + "value." + e.Context().String() + ":" + e.Description()
+		}
+		fieldErrors = append(fieldErrors, errorDesc)
+	}
+
+	return errors.New(strings.Join(fieldErrors[:], "/ "))
+}
+
+func (db *Database) ValidateAppMetrics(appMetrics []AppMetric) (err error) {
+	var calculatedErrors []AppMetricValidationError
+	for _, metric := range appMetrics {
+		schema := EventSchemaMap[metric.Event]
+
+		if schema == nil {
+			return errors.New("No schema defined for: " + string(metric.Event))
+		}
+
+		schemaLoader := gojsonschema.NewGoLoader(schema)
+		valLoader := gojsonschema.NewStringLoader(string(metric.Value))
+		res, err := gojsonschema.Validate(schemaLoader, valLoader)
+
+		if err != nil {
+			return err
+		}
+
+		// validate all metrics and save errors
+		if !res.Valid() {
+			calculatedErrors = append(calculatedErrors, AppMetricValidationError{metric, res.Errors()})
+		}
+	}
+
+	if len(calculatedErrors) > 0 {
+		return jsonschemaErrorsToError(calculatedErrors)
+	}
+	return
+}
+
+func (db *Database) SaveAppMetrics(appMetrics []AppMetric) (err error) {
+	var (
+		tx     *sql.Tx
+		insert *sql.Stmt
+	)
+
+	// make sure that the shape of the metric is same as expected
+	err = db.ValidateAppMetrics(appMetrics)
+	if err != nil {
+		return err
+	}
+
+	// start txn
+	tx, err = db.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+			return
+		}
+		_ = tx.Rollback()
+	}()
+
+	insert, err = tx.Prepare("INSERT INTO app_metrics (event, value, app_version, operating_system) VALUES (?, ?, ?, ?)")
+	if err != nil {
+		return err
+	}
+
+	for _, metric := range appMetrics {
+		_, err := insert.Exec(metric.Event, metric.Value, metric.AppVersion, metric.OS)
```

Comment:
I think here you might be shadowing `err`, and what's going to happen is that fi `err` on 135 is not nil, it will still try to commit the transaction, because `err` (the one declared at the beginning of the function and in the `defer` block is nil).

So you likely want to just do:
`_, err = insert.Exec...`

Reply:
Not sure if this is what you suggested, but if I get rid of the explicit `return`, the linter complains that the `err` is defined but never used:

<img width="970" alt="Screenshot 2021-03-16 at 4 58 16 PM" src="https://user-images.githubusercontent.com/1925158/111302086-f801d800-8678-11eb-824b-d9a25f5aee77.png">

Reply:
yes, you also need to remove the initialization,
`:=` should be `=`, otherwise you will be shadowing `err`